### PR TITLE
Webpack TS Configuration Initial Implementation

### PR DESCRIPTION
**Pull request by AlbertoBrusa**
**Background**
Webpack now supports configuration written in TypeScript.
Currently, modular-scripts has two folders of JS code copied over from CRA, adapted as needed, to create the Webpack configuration for Modular. Modular's build and start script run JS scripts with execa() to start and build with Webpack.
Problems: 
- JavaScript Code outside the /src/ directory - hard to maintain and work on 
- Duplicated code - some of these scripts have been translated to TS and placed in /src/ as they are required by other parts of Modular 
- Running the scripts with execa() means having to pass all variables as Environment variables

**This PR**
We now create the Webpack configuration in TypeScript, and run the build and start scripts as functions, passing variables directly.
- No more internal Env variables 
- TypeChecking in Webpack configuration 
- Moved all scripts into /src/build-scripts where we can manage them better
- react-scripts and react-dev-utils are now integrated in the modular build logic under build-scripts/webpack-scripts/ and match what esbuild-scripts do/how they are called
- Duplicated logic from esbuild and webpack now deduped and placed in common-scripts

To Do: 

- [x] Fix CDN rewriting cause it seems very much broken right now (esmView.test.ts)
- [x] Fix conflicts with https://github.com/jpmorganchase/modular/pull/2282
- [x] Improve folder structure 
- [x] Refactor/rename/cleanup 

There's still a lot of work to be done in future PRs:
- Still some duplicated pieces of code
- Many scripts still in JS
    - The script that provides the baseConfig for Webpack has a non-typed dependency, and in turn it uses a lot of scripts that have to stay written in JS 

**Few notes for those reviewing this PR**:
Things have been moved around so much that a lot of code appears as 'new', while in fact it's just moved/typed but using existing logic. I've held back from making actual functional code changes to existing scripts unless absolutely necessary.

### CODE REVIEW

File: packages/modular-scripts/src/build/index.ts [hunk 1]
```diff
@@ -135,35 +133,25 @@ async function buildStandalone(
...
 
-    const buildScript = require.resolve(
-      'modular-scripts/react-scripts/scripts/build.js',
-    );
+    // If it's an app, set it at ESBUILD_TARGET_FACTORY or default to es2015
+    // If it's not an app it's an ESM view, then we need es2020
+    const esbuildTargetFactory = isApp ? browserTarget : ['es2020'];
+    let targetPath = await getWorkspaceLocation(target);
 
-    // TODO: this shouldn't be sync
```
Comment (AlbertoBrusa) [inline packages/modular-scripts/src/build/index.ts:142]: We no longer run the .js file separately with ENV variables, instead calling a simple buildWebpack() function, passing it all the necessary variables.

File: packages/modular-scripts/src/build/index.ts [hunk 2]
```diff
@@ -135,35 +133,25 @@ async function buildStandalone(
...
-        MODULAR_IS_APP: JSON.stringify(isApp),
-        MODULAR_IMPORT_MAP: JSON.stringify(Object.fromEntries(importMap || [])),
-        MODULAR_USE_REACT_CREATE_ROOT: JSON.stringify(useReactCreateRoot),
-        INTERNAL_PUBLIC_URL: getConfig('publicUrl', targetDirectory),
-        INTERNAL_GENERATE_SOURCEMAP: String(
-          getConfig('generateSourceMap', targetDirectory),
-        ),
-      },
-    });
+    await buildWebpack(
```
Comment (AlbertoBrusa) [inline packages/modular-scripts/src/build/index.ts:141]: Here we're using the new TS function instead of using exec() to run a JS script

File: packages/modular-scripts/src/esbuild-scripts/start/index.ts [hunk 3]
```diff
@@ -27,7 +27,7 @@ import * as logger from '../../utils/logger';
 import prepareUrls, { InstructionURLS } from '../config/urls';
 import { createIndex, indexFile, createViewTrampoline } from '../api';
 import createEsbuildConfig from '../config/createEsbuildConfig';
-import createLaunchEditorMiddleware from '../../../react-dev-utils/errorOverlayMiddleware.js';
+import createLaunchEditorMiddleware from '../../react-dev-utils/errorOverlayMiddleware.js';
```
Comment (AlbertoBrusa) [inline packages/modular-scripts/src/esbuild-scripts/start/index.ts:30]: From moving all of react-scripts and react-dev-utils into modular /src/

File: packages/modular-scripts/src/react-scripts/config/getHttpsConfig.js [hunk 4]
```diff
@@ -4,7 +4,6 @@ const fs = require('fs');
 const path = require('path');
 const crypto = require('crypto');
 const chalk = require('chalk');
-const paths = require('./paths');
```
Comment (AlbertoBrusa) [inline packages/modular-scripts/src/react-scripts/config/getHttpsConfig.js:7]: Paths are no longer a static js file but a TS script that's called only once per build and who's output is passed down to child functions that need it

File: packages/modular-scripts/src/react-scripts/config/modules.ts [hunk 5]
```diff
@@ -1,23 +1,24 @@
...
+import { Paths } from '../../utils/createPaths';
 
 /**
  * Get additional module paths based on the baseUrl of a compilerOptions object.
  *
  * @param {Object} options
  */
-function getAdditionalModulePaths(options = {}) {
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function getAdditionalModulePaths(options: any = {}, paths: Paths) {
```
Comment (AlbertoBrusa) [inline packages/modular-scripts/src/react-scripts/config/modules.ts:13]: TODO: Type more accurately

File: packages/modular-scripts/src/react-scripts/config/modules.ts [hunk 6]
```diff
@@ -54,13 +55,16 @@ function getAdditionalModulePaths(options = {}) {
  *
  * @param {*} options
  */
-function getWebpackAliases(options = {}) {
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function getWebpackAliases(options: any = {}, paths: Paths) {
```
Comment (AlbertoBrusa) [inline packages/modular-scripts/src/react-scripts/config/modules.ts:59]: TODO: Type more accurately

File: packages/modular-scripts/src/react-scripts/config/webpack.config.ts [hunk 7]
```diff
@@ -72,36 +57,41 @@ module.exports = function (webpackEnv) {
     dependencyMap,
   });
 
+  const performanceConfiguration: Configuration = {
```
Comment (AlbertoBrusa) [inline packages/modular-scripts/src/react-scripts/config/webpack.config.ts:60]: Extracted from baseConfig to here for typing issues

File: packages/modular-scripts/src/react-scripts/config/parts/pluginConfig.ts [hunk 8]
```diff
@@ -106,7 +84,6 @@ module.exports = function createPluginConfig({
         new ForkTsCheckerWebpackPlugin({
           async: isEnvDevelopment,
           typescript: {
-            async: isEnvDevelopment,
```
Comment (AlbertoBrusa) [inline packages/modular-scripts/src/react-scripts/config/parts/pluginConfig.ts:109]: webpack config type claims async isn't a property of typescript {}

File: packages/modular-scripts/package.json [hunk 9]
```diff
@@ -125,11 +129,12 @@
     "tmp": "^0.2.1",
     "ts-jest": "^29.0.0",
     "ts-morph": "^14.0.0",
+    "ts-node": "^10.9.1",
```
Comment (AlbertoBrusa) [inline packages/modular-scripts/package.json:132]: Needed for TS webpack config

File: packages/modular-scripts/src/build/index.ts [hunk 10]
```diff
@@ -135,35 +133,25 @@ async function buildStandalone(
     // create-react-app doesn't support plain module outputs yet,
     // so --preserve-modules has no effect here
 
-    const buildScript = require.resolve(
-      'modular-scripts/react-scripts/scripts/build.js',
-    );
+    // If it's an app, set it at ESBUILD_TARGET_FACTORY or default to es2015
+    // If it's not an app it's an ESM view, then we need es2020
+    const esbuildTargetFactory = isApp ? browserTarget : ['es2020'];
```
Comment (AlbertoBrusa) [inline packages/modular-scripts/src/build/index.ts:138]: Casual observation:
Webpack build logic uses 'isApp' a lot, ie if not app, then build as if it was an ESM view etc
This would potentially cause problems/not be ideal if adding more types - I feel like we should be doing every time things like if (type === app) else if (type === esmView) else Error Type not supported or similar, and handle every type we have so far - 
unless we decide its absolutely impossible another type will ever go through the build process

File: .github/workflows/test.yml [hunk 11]
```diff
@@ -5,9 +5,9 @@ name: Test
 
 on:
   push:
-    branches: [main, 'release/**', 'feature/v4.2']
+    branches: [main, 'release/**', 'feature/**']
   pull_request:
-    branches: [main, 'release/**', 'feature/v4.2']
+    branches: [main, 'release/**', 'feature/**']
```
Comment (AlbertoBrusa) [inline .github/workflows/test.yml:10]: Any reason not to do this if we keep up with the 'feature' branch naming convention?

File: packages/modular-scripts/src/react-scripts/config/parts/pluginConfig.ts (hunk 8 above)
Comment (AlbertoBrusa) [reply packages/modular-scripts/src/react-scripts/config/parts/pluginConfig.ts:109]: Allowed properties of typescript:
```
    memoryLimit?: number;
    configFile?: string;
    configOverwrite?: TypeScriptConfigOverwrite;
    context?: string;
    build?: boolean;
    mode?: 'readonly' | 'write-tsbuildinfo' | 'write-dts' | 'write-references';
    diagnosticOptions?: Partial<TypeScriptDiagnosticsOptions>;
    extensions?: {
        vue?: TypeScriptVueExtensionOptions;
    };
    profile?: boolean;
    typescriptPath?: string;
```

I think we can put this down as a mistake until now uncaught because of lack of typings

File: packages/modular-scripts/src/react-scripts/scripts/startWebpack.ts [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+import isCI from 'is-ci';
+
+import getConfig from '../config/webpack.config';
+import createDevServerConfig from '../config/webpackDevServer.config';
+import { Paths } from '../../utils/determineTargetPaths';
+
+// Makes the script crash on unhandled rejections instead of silently
+// ignoring them. In the future, promise rejections that are not handled will
+// terminate the Node.js process with a non-zero exit code.
+process.on('unhandledRejection', (err) => {
```
Comment (AlbertoBrusa) [inline packages/modular-scripts/src/react-scripts/scripts/startWebpack.ts:22]: Does this still make sense now that we're not running it as a separate JS script anymore

File: packages/modular-scripts/src/build-scripts/webpack-scripts/utils/logger.js [hunk 13]
```diff
@@ -1,5 +1,6 @@
 'use strict';
 
+// TODO: Duplicate of logger.ts in utils, remove once all JS scripts that use this file have been typed
```
Comment (AlbertoBrusa) [inline packages/modular-scripts/src/build-scripts/webpack-scripts/utils/logger.js:3]: This will need to be a separate PR as it's another significant bit of work

Comment (AlbertoBrusa) [reply packages/modular-scripts/src/build-scripts/webpack-scripts/utils/logger.js]: Ended up doing as part of this PR

File: packages/modular-scripts/src/build-scripts/webpack-scripts/startWebpack.ts [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+        proxyConfig,
+        urls.lanUrlForConfig,
+        paths,
+      );
+
+      // Types from WebpackDevServer and Webpack overlap but unfortunately aren't compatible
+      // even though when you look into it Webpack-Dev-Server simply imports the Webpack types
+      const devServer = new WebpackDevServer(
+        serverConfig,
+        compiler as unknown as Compiler,
```
Comment (AlbertoBrusa) [inline packages/modular-scripts/src/build-scripts/webpack-scripts/startWebpack.ts:102]: TypeScript doesn't manage to figure out that the Compiler Type from Webpack and that from Webpack-Dev-Server are the same, and fails to compare the two as they are 'too big' to compare and assert that they are the same.

**Comment by cristiano-belloni** [comment]
@AlbertoBrusa I think we need a minor changeset here

File: packages/modular-scripts/package.json [hunk 15]
```diff
@@ -125,11 +131,12 @@
...
```
Comment (cristiano-belloni) [inline packages/modular-scripts/package.json:134]: I think we already have this in the root pkg, no?

File: packages/modular-scripts/package.json [hunk 16]
```diff
@@ -152,8 +159,8 @@
   "devDependencies": {
     "@modular-scripts/modular-types": "1.2.0",
     "@schemastore/package": "0.0.6",
-    "@schemastore/tsconfig": "0.0.9",
```
Comment (cristiano-belloni) [inline packages/modular-scripts/package.json:155]: Why is this removed?

Comment (AlbertoBrusa) [reply packages/modular-scripts/package.json:155]: Not sure, must have been unintentional - any reason we have duplicate dependencies in devDependencies and normal dependencies?

File: packages/modular-scripts/src/build-scripts/build-package/index.ts [hunk 17]
```diff
@@ -82,7 +82,7 @@ export async function buildPackage(
     );
   });
 
-  await maybeCopyRootLicense(target, targetOutputDirectory);
+  maybeCopyRootLicense(target, targetOutputDirectory);
```
Comment (cristiano-belloni) [inline packages/modular-scripts/src/build-scripts/build-package/index.ts:85]: Lol was this actually sync?

File: packages/modular-scripts/src/build-scripts/buildStandalone.ts [hunk 18]
```diff
@@ -105,78 +101,58 @@ export async function buildStandalone(
...
+    const stats = await buildWebpack(
+      esbuildTargetFactory,
+      isApp,
+      importMap,
+      useReactCreateRoot,
+      styleImports,
+      paths,
     );
 
-    // TODO: this shouldn't be sync
```
Comment (cristiano-belloni) [inline packages/modular-scripts/src/build-scripts/buildStandalone.ts:131]: So beautiful to get rid of this horror!

File: packages/modular-scripts/src/build-scripts/buildStandalone.ts [hunk 19]
```diff
@@ -199,9 +175,9 @@ export async function buildStandalone(
   // Add dependencies from source and bundled dependencies to target package.json
   const targetPackageJson = (await fs.readJSON(
     path.join(targetDirectory, 'package.json'),
-  )) as CoreProperties;
+  )) as ModularPackageJson;
   // Copy selected fields of package.json over
-  fs.writeJSON(
+  fs.writeJsonSync(
```
Comment (cristiano-belloni) [inline packages/modular-scripts/src/build-scripts/buildStandalone.ts:180]: Why not `await fs.writeJSON`?

File: packages/modular-scripts/src/build-scripts/webpack-scripts/config/getHttpsConfig.ts [hunk 20]
```diff
@@ -1,38 +1,52 @@
...
-    );
-  }
 
-  try {
-    // privateDecrypt will throw an error with an invalid key
-    crypto.privateDecrypt(key, encrypted);
+    try {
+      // privateDecrypt will throw an error with an invalid key
+      crypto.privateDecrypt(key, encrypted);
+    } catch (err) {
```
Comment (cristiano-belloni) [inline packages/modular-scripts/src/build-scripts/webpack-scripts/config/getHttpsConfig.ts:28]: Are those two consecutive `catch`es duplicated?

File: packages/modular-scripts/src/build-scripts/buildStandalone.ts (hunk 18 above)
Comment (AlbertoBrusa) [reply packages/modular-scripts/src/build-scripts/buildStandalone.ts:131]: Agreed :)

File: packages/modular-scripts/src/build-scripts/buildStandalone.ts (hunk 19 above)
Comment (AlbertoBrusa) [reply packages/modular-scripts/src/build-scripts/buildStandalone.ts:180]: Is there a benefit to one over the other? 
Not sure why I changed it to that instead of awaiting

File: packages/modular-scripts/src/build-scripts/webpack-scripts/config/getHttpsConfig.ts [hunk 21]
```diff
@@ -1,38 +1,52 @@
...
-const fs = require('fs');
-const path = require('path');
-const crypto = require('crypto');
-const chalk = require('chalk');
-const paths = require('./paths');
+import fs from 'fs';
+import path from 'path';
+import crypto from 'crypto';
+import chalk from 'chalk';
+import { Paths } from '../../common-scripts/determineTargetPaths';
```
Comment (cristiano-belloni) [inline packages/modular-scripts/src/build-scripts/webpack-scripts/config/getHttpsConfig.ts:5]: This looks like it should be an `import type`

File: packages/modular-scripts/src/build-scripts/webpack-scripts/config/modules.ts [hunk 22]
```diff
@@ -1,17 +1,19 @@
...
-const fs = require('fs');
-const path = require('path');
-const paths = require('./paths');
-const chalk = require('chalk');
-const resolve = require('resolve');
+import fs from 'fs';
+import path from 'path';
+import chalk from 'chalk';
+import resolve from 'resolve';
+import { Paths } from '../../common-scripts/determineTargetPaths';
```
Comment (cristiano-belloni) [inline packages/modular-scripts/src/build-scripts/webpack-scripts/config/modules.ts:5]: I think this is also an `import type`

File: packages/modular-scripts/src/build-scripts/webpack-scripts/config/parts/appConfig.ts [hunk 23]
```diff
@@ -51,21 +54,24 @@ function createPluginConfig({
               }
             : undefined,
         ),
-      ),
+      ) as unknown as WebpackPluginInstance,
```
Comment (cristiano-belloni) [inline packages/modular-scripts/src/build-scripts/webpack-scripts/config/parts/appConfig.ts:57]: argh, why do we need this?

Comment (AlbertoBrusa) [reply packages/modular-scripts/src/build-scripts/webpack-scripts/config/parts/appConfig.ts:57]: TypeScript error:
`Excessive stack depth comparing types 'HtmlWebpackPlugin' and 'WebpackPluginInstance'.ts(2321)`
Essentially it doesn't manage to check that the two are actually compatible

Comment (AlbertoBrusa) [reply packages/modular-scripts/src/build-scripts/webpack-scripts/config/parts/appConfig.ts:57]: Theres a few instances of this unfortunately

File: packages/modular-scripts/src/build-scripts/webpack-scripts/config/parts/esmViewConfig.ts [hunk 24]
```diff
@@ -1,18 +1,21 @@
...
+import HtmlWebpackPlugin from 'html-webpack-plugin';
+import dedent from 'dedent';
+import path from 'path';
+import webpack, { Configuration } from 'webpack';
+import { parsePackageName } from '../../../../utils/parsePackageName';
+import { rewriteModuleSpecifier } from '../../../../utils/buildImportMap';
+import fs from 'fs-extra';
+import parse5 from 'parse5';
+import { Element } from 'parse5/dist/tree-adapters/default';
+import { Paths } from '../../../common-scripts/determineTargetPaths';
```
Comment (cristiano-belloni) [inline packages/modular-scripts/src/build-scripts/webpack-scripts/config/parts/esmViewConfig.ts:10]: This two might be `import type`

File: packages/modular-scripts/src/build-scripts/webpack-scripts/config/parts/esmViewConfig.ts [hunk 25]
```diff
@@ -1,18 +1,21 @@
...
+import HtmlWebpackPlugin from 'html-webpack-plugin';
+import dedent from 'dedent';
+import path from 'path';
+import webpack, { Configuration } from 'webpack';
+import { parsePackageName } from '../../../../utils/parsePackageName';
+import { rewriteModuleSpecifier } from '../../../../utils/buildImportMap';
+import fs from 'fs-extra';
+import parse5 from 'parse5';
+import { Element } from 'parse5/dist/tree-adapters/default';
+import type { Paths } from '../../../common-scripts/determineTargetPaths';
```
Comment (cristiano-belloni) [inline packages/modular-scripts/src/build-scripts/webpack-scripts/config/parts/esmViewConfig.ts:10]: `import type`, this line and the one below?

File: packages/modular-scripts/src/build-scripts/buildStandalone.ts (hunk 19 above)
Comment (cristiano-belloni) [reply packages/modular-scripts/src/build-scripts/buildStandalone.ts:180]: There is when we enable parallel builds, because async calls can be parallelised by Node, sync calls block the main thread until they return, so they're always serialized.

File: packages/modular-scripts/src/build-scripts/webpack-scripts/config/parts/esmViewConfig.ts (hunk 24 above)
Comment (AlbertoBrusa) [reply packages/modular-scripts/src/build-scripts/webpack-scripts/config/parts/esmViewConfig.ts]: WebpackConfiguration and Configuration are the same from different libraries (webpack and webpack-dev-server)
will dedup

File: packages/modular-scripts/src/build-scripts/common-scripts/determineTargetPaths.ts [hunk 26]
```diff
@@ -98,12 +127,17 @@ export default async function createPaths(target: string): Promise<Paths> {
   const appPackageJson = resolveApp('package.json');
   const appSrc = resolveApp('src');
   const appTsConfig = resolveApp('tsconfig.json');
+  const appJsConfig = resolveApp('jsconfig.json');
```
[the remark this answers is not in the document]
Comment (cristiano-belloni) [reply packages/modular-scripts/src/build-scripts/common-scripts/determineTargetPaths.ts:130]: Happy with that. Never used it in my life, but evidently there's someone who uses it.

File: packages/modular-scripts/src/build-scripts/webpack-scripts/config/parts/appConfig.ts (hunk 23 above)
Comment (cristiano-belloni) [reply packages/modular-scripts/src/build-scripts/webpack-scripts/config/parts/appConfig.ts:57]: Understood, can we explain it with a comment specifying `... this happens with current Typescript version <x.y.z>`? I foresee everyone coming here spending a couple of hours trying to fix this 🤣

File: packages/modular-scripts/src/build-scripts/webpack-scripts/config/parts/productionConfig.ts [hunk 27]
```diff
@@ -1,8 +1,13 @@
-'use strict';
-const path = require('path');
-const MiniCssExtractPlugin = require('mini-css-extract-plugin');
+import path from 'path';
+import MiniCssExtractPlugin from 'mini-css-extract-plugin';
+import type { Paths } from '../../../common-scripts/determineTargetPaths';
+import { WebpackConfiguration } from 'webpack-dev-server';
```
Comment (cristiano-belloni) [inline packages/modular-scripts/src/build-scripts/webpack-scripts/config/parts/productionConfig.ts:4]: `import types` (above & below too)

File: packages/modular-scripts/src/build-scripts/webpack-scripts/config/webpack.config.ts [hunk 28]
```diff
@@ -0,0 +1,168 @@
...
+const imageInlineSizeLimit = parseInt(
+  process.env.IMAGE_INLINE_SIZE_LIMIT || '10000',
+);
+
+/**
+ * Generate Webpack Configuration
+ * This is the production and development configuration.
+ * It is focused on developer experience, fast rebuilds, and a minimal bundle.
+ * @param isEnvProduction True when building, false when just starting
+ * @param esbuildTargetFactory
```
Comment (cristiano-belloni) [inline packages/modular-scripts/src/build-scripts/webpack-scripts/config/webpack.config.ts:28]: We should probably explain what these parameters are in the jsdoc

File: packages/modular-scripts/src/build-scripts/webpack-scripts/config/webpackDevServer.config.ts [hunk 29]
```diff
@@ -1,20 +1,24 @@
...
-const paths = require('./paths');
-const getHttpsConfig = require('./getHttpsConfig');
+import fs from 'fs';
+import errorOverlayMiddleware from '../../common-scripts/errorOverlayMiddleware';
+import evalSourceMapMiddleware from '../utils/evalSourceMapMiddleware';
+import noopServiceWorkerMiddleware from '../utils/noopServiceWorkerMiddleware';
+import ignoredFiles from '../utils/ignoredFiles';
+import redirectServedPath from '../utils/redirectServedPathMiddleware';
+import getHttpsConfig from './getHttpsConfig';
+import { Configuration, ProxyConfigArray } from 'webpack-dev-server';
```
Comment (cristiano-belloni) [inline packages/modular-scripts/src/build-scripts/webpack-scripts/config/webpackDevServer.config.ts:8]: Another `import type`

File: packages/modular-scripts/src/start.ts [hunk 30]
```diff
@@ -57,14 +55,16 @@ async function start(packageName: string): Promise<void> {
     );
   }
 
+  let paths;
   const isEsmView = isModularType(targetPath, 'esm-view');
   const isView = isModularType(targetPath, 'view');
   if (isView) {
     targetPath = stageView(target);
+    paths = determineTargetPaths(target, targetPath);
```
Comment (cristiano-belloni) [inline packages/modular-scripts/src/start.ts:63]: `paths = determineTargetPaths(target, targetPath);` in the if branch is repeated in the else branch. Maybe it just needs to be outside the if?

Comment (AlbertoBrusa) [reply packages/modular-scripts/src/start.ts:63]: Target is updated in one of the two if branches, thus changing the output of determine target branches 
It was actually a bug that I took a long time to figure out

File: packages/modular-scripts/src/build-scripts/build-package/index.ts (hunk 17 above)
Comment (AlbertoBrusa) [reply packages/modular-scripts/src/build-scripts/build-package/index.ts:85]: yep

File: packages/modular-scripts/src/start.ts (hunk 30 above)
Comment (cristiano-belloni) [reply packages/modular-scripts/src/start.ts:63]: I see. We'll need to disentangle this code, but not for now. Can you put a `//TODO: this needs to be rewritten in a clearer way` or something equivalent?

Comment (AlbertoBrusa) [reply packages/modular-scripts/src/start.ts]: Now that I've revisited it I cleaned it up

File: packages/modular-scripts/src/build-scripts/common-scripts/determineTargetPaths.ts [hunk 31]
```diff
@@ -90,6 +99,26 @@ export default async function createPaths(target: string): Promise<Paths> {
...
+    require.resolve(resolveModular('package.json')),
+  ) as { workspaces: string[] | { packages: string[] } };
+  const workspaceDefinitions =
+    (Array.isArray(rootManifest?.workspaces)
+      ? rootManifest?.workspaces
+      : rootManifest?.workspaces?.packages) || [];
+
+  // Calculate all the possible workspace directories. We need to convert paths to posix separator to feed it into globby
+  // and convert back to native separator after
+  const workspaceDirectories = globby
```
[the remark this answers is not in the document]
Comment (cristiano-belloni) [reply packages/modular-scripts/src/build-scripts/common-scripts/determineTargetPaths.ts:116]: This needs more exploration, let's not change the logic for now.

**Review by cristiano-belloni — APPROVED** [review]
